### PR TITLE
[AArch64] Add soft-float ABI

### PR DESCRIPTION
**Pull request by ostannard**
This is re-working of #74460, which adds a soft-float ABI for AArch64. That was reverted because it causes errors when building the linux and fuchsia kernels.

The problem is that GCC's implementation of the ABI compatibility checks when using the hard-float ABI on a target without FP registers does it's checks after optimisation. The previous version of this patch reported errors for all uses of floating-point types, which is stricter than what GCC does in practice.

This changes two things compared to the first version:
* Only check the types of function arguments and returns, not the types of other values. This is more relaxed than GCC, while still guaranteeing ABI compatibility.
* Move the check from Sema to CodeGen, so that inline functions are only checked if they are actually used. There are some cases in the linux kernel which depend on this behaviour of GCC.

I've tested this by building the linux kernel, but don't have a fuchsia build set up. @Prabhuk could you check this?

### CODE REVIEW

**Comment by ostannard** [comment]
Ping

**Comment by Prabhuk** [comment]
Apologies that this had slipped out of my radar. I am trying this patch now agains the reproducer from the original failure. I'll update the results here in a few hours. Thank you.

**Comment by Prabhuk** [comment]
Verified that the reproducer for the original crash fails before the revert patch: https://github.com/llvm/llvm-project/pull/82032/commits/13bf4429ba2e33f9069f632c974d8fdf121acefa

and toolchain built with current patch *does not fail* against the reproducer.

Though I haven't tried a full Fuchsia Kernel build yet, I believe the test against the reproducer suggests this patch is safe to land. Thanks @ostannard.

**Review by Prabhuk — APPROVED** [review]

**Comment by shafik** [comment]
This is linked to the following regression: https://github.com/llvm/llvm-project/issues/111293

Please investigate.

File: clang/lib/CodeGen/ModuleBuilder.cpp [hunk 1]
```diff
@@ -180,7 +180,7 @@ namespace {
 
     bool HandleTopLevelDecl(DeclGroupRef DG) override {
       // FIXME: Why not return false and abort parsing?
-      if (Diags.hasErrorOccurred())
+      if (Diags.hasUnrecoverableErrorOccurred())
```
Comment (shafik) [inline clang/lib/CodeGen/ModuleBuilder.cpp:183]: Is this really a reasonable change? Doesn't that mean will be codegen'ing on erroneous code?

**Comment by shafik** [comment]
This is also linked to this clang-19 regression: https://github.com/llvm/llvm-project/issues/110914

**Comment by shafik** [comment]
ping

**Comment by ostannard** [comment]
Those two regressions turned out to be an existing bug, which previously only triggered with `-Wno-int-conversion`. They should be fixed by #114217 and #114218.

File: clang/lib/CodeGen/ModuleBuilder.cpp (hunk 1 above)
Comment (ostannard) [reply clang/lib/CodeGen/ModuleBuilder.cpp:183]: I think this is reasonable, it allows codegen to continue if all of the errors which have occurred are downgradable, which happens before this change if the correct `-Wno-whatever` or `-Wno-error=whatever` option is used. We still bail out here if any of the errors are unrecoverable.

**Pull request merged**